### ISSUE TITLE
MultiheadAttention and DDP incompatability

### ISSUE THREAD

**Issue by roderickObrist** [pytorch/pytorch#26698]
Hi Guys,

First of all thanks for a great code base. I know I'm not following the bug guidelines but I thought you guys might appreciate a heads up.

nn.MultiheadAttention allows you to provide different keys/value dimensions. When this happens the flag _qkv_same_embed_dim becomes False.

``` 
       # MultiheadAttention constructor activation.py:690
        self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))

        if self._qkv_same_embed_dim is False:
            self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
            self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
            self.v_proj_weight = Parameter(torch.Tensor(embed_dim, self.vdim))
```

As you can see self.in_proj_weight is initialised regardless.

In DDP all parameters expect gradients, and thus are expected to be used when calculating loss. This will throw a RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one.....

If you would like to reproduce this bug you must put a MultiheadAttention with different query/key/value dimensions and wrap it in DDP.

To fix the bug:

``` 
       # MultiheadAttention constructor activation.py:690
        if self._qkv_same_embed_dim is False:
            self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
            self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
            self.v_proj_weight = Parameter(torch.Tensor(embed_dim, self.vdim))
            self.register_parameter('in_proj_weight', None)
        else:
            self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
```

**Comment by soumith** [comment]
cc: @zhangguanheng66

**Comment by zhangguanheng66** [comment]
@roderickObrist Thanks for bringing up the issue. I will reproduce the error and fix it.

**Comment by zhangguanheng66** [comment]
@roderickObrist I try the following script and cannot reproduce the issue. Wrap up `nn.MultiheadAttention` within `DDP`.

```
import os
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
import torch.multiprocessing as mp
from torch.nn.parallel import DistributedDataParallel as DDP


def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '12355'

    # initialize the process group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

    # Explicitly setting seed to make sure that models created in two processes
    # start from same random weights and biases.
    torch.manual_seed(42)


def cleanup():
    dist.destroy_process_group()


def demo_basic(rank, world_size):
    setup(rank, world_size)

    # setup devices for this process, rank 1 uses GPUs [0, 1, 2, 3] and
    # rank 2 uses GPUs [4, 5, 6, 7].
    n = torch.cuda.device_count() // world_size
    device_ids = list(range(rank * n, (rank + 1) * n))

    # create model and move it to device_ids[0]
    embed_dim = 128
    kv_embed_dim = 1024
    num_heads = 8
    model = torch.nn.MultiheadAttention(embed_dim, num_heads,
                                        kdim=kv_embed_dim,
                                        vdim=kv_embed_dim).to(device_ids[0])

    # output_device defaults to device_ids[0]
    ddp_model = DDP(model, device_ids=device_ids)

    # inputs
    src_len, tgt_len, batch_size = 10, 5, 16
    query = torch.rand((tgt_len, batch_size, embed_dim)).to(device_ids[0])
    key = torch.rand((src_len, batch_size, kv_embed_dim)).to(device_ids[0])
    value = key.to(device_ids[0])

    outputs = ddp_model(query, key, value)

    cleanup()


def run_demo(demo_fn, world_size):
    mp.spawn(demo_fn,
             args=(world_size,),
             nprocs=world_size,
             join=True)


if __name__ == "__main__":
    run_demo(demo_basic, 2)
```

**Comment by roderickObrist** [comment]
@zhangguanheng66 Hi.

So that error is only triggered on the second forward pass after a backward().

Here is a demo_basic that should trigger it.

```

def demo_basic(rank, world_size):
    setup(rank, world_size)

    # setup devices for this process, rank 1 uses GPUs [0, 1, 2, 3] and
    # rank 2 uses GPUs [4, 5, 6, 7].
    n = torch.cuda.device_count() // world_size
    device_ids = list(range(rank * n, (rank + 1) * n))

    # create model and move it to device_ids[0]
    embed_dim = 128
    kv_embed_dim = 1024
    num_heads = 8
    model = torch.nn.MultiheadAttention(embed_dim, num_heads,
                                        kdim=kv_embed_dim,
                                        vdim=kv_embed_dim)

    # output_device defaults to device_ids[0]
    ddp_model = DDP(model, device_ids=device_ids)

    # inputs
    src_len, tgt_len, batch_size = 10, 5, 16
    query = torch.rand((tgt_len, batch_size, embed_dim))
    key = torch.rand((src_len, batch_size, kv_embed_dim))
    value = key

    outputs = ddp_model(query, key, value)[0]
    loss_fn = torch.nn.functional.l1_loss

    loss = loss_fn(outputs, torch.zeros_like(outputs))

    loss.backward()
    ddp_model.zero_grad()

    outputs = ddp_model(query, key, value)[0]

    cleanup()

```

That function should work with embed_dim == kv_embed_dim, but should fail without.
I took the GPU calls out because I don't have cuda on my desktop, but you don't have to. You can just add:

```
    outputs = ddp_model(query, key, value)[0]
    loss_fn = torch.nn.functional.l1_loss

    loss = loss_fn(outputs, torch.zeros_like(outputs))

    loss.backward()
    ddp_model.zero_grad()

    outputs = ddp_model(query, key, value)[0]
```

Which is your standard training flow.

Also, while you don't have to register the parameter as none, that attribute is read later. So you can either change the attribute read in the forward function, or change the __init__ and it will run without error. But that's up to you guys. 

Also I'd say it's a MultiHeadAttention error more than a DDP error.

Long story short:

- for DDP compatibility all all model parameters must have gradients after the loss is propagated.
- When MultiHeadAttention is initialised with different kv and embed dim then in_proj_weight is still registered but not used

**Comment by zhangguanheng66** [comment]
@mrshenli I have discussed with Shen offline. I will submit a PR to fix it soon.

**Comment by zhangguanheng66** [comment]
@roderickObrist After talking with our DDP people, it seems that the best solution for now is to add the `find_unused_parameters` parameter in `DistributedDataParallel` module.

**Comment by roderickObrist** [comment]
@zhangguanheng66 That was the first thing I tried, when I was tinkering and finding what the problem is.

But It was throwing another really big error, which I can't seem to reproduce at the moment.

I'm not in need of help by the way, I patched my own version of it before I filed the bug.

I just thought you guys might like to know.

**Comment by zhangguanheng66** [comment]
> @zhangguanheng66 That was the first thing I tried, when I was tinkering and finding what the problem is.
> 
> But It was throwing another really big error, which I can't seem to reproduce at the moment.
> 
> I'm not in need of help by the way, I patched my own version of it before I filed the bug.
> 
> I just thought you guys might like to know.


@roderickObrist Thanks for letting us know this issue. I would like to know the error when using `find_unused_parameters`, because it works on my side. Once you can reproduce the error, we can resume the work from there.

**Comment by zhangguanheng66** [comment]
Close the issue. A fix has been landed [26826](https://github.com/pytorch/pytorch/pull/26826)
